### PR TITLE
This table combine the ethnicity and gender of the donors

### PR DESCRIPTION
**Pull request by Roselkisma**
I wrote the labels manually, I don't know if there is another way to do it. #1932

### CODE REVIEW

File: context/package.json [hunk 1]
```diff
@@ -30,6 +31,7 @@
     "pretty-bytes": "^5.3.0",
     "prop-types": "^15.7.2",
     "react": "^16.12.0",
+    "react-chartjs-2": "^3.0.3",
```
Comment (mccalluc) [inline context/package.json:34]:
```suggestion
```

File: context/package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "@visx/tooltip": "^1.7.2",
     "ajv": "^6.12.3",
     "bowser": "^2.11.0",
+    "chart.js": "^3.4.1",
```
Comment (mccalluc) [inline context/package.json:21]:
```suggestion
```

File: context/package-lock.json [hunk 3]
```diff
@@ -30499,6 +30504,14 @@
...
       }
     },
+    "react-chartjs-2": {
+      "version": "3.0.3",
+      "resolved": "https://registry.npmjs.org/react-chartjs-2/-/react-chartjs-2-3.0.3.tgz",
+      "integrity": "sha512-jOFZKwZ8sMLkddewZ/tToxuu4pYimAvvY5I6uK+hCpSFT16Pvo2bdHhUoZ0X87zu9I+dx2I+JCqaLN6XhmrbDg==",
+      "requires": {
+        "lodash": "^4.17.19"
+      }
+    },
```
Comment (mccalluc) [inline context/package-lock.json:30514]:
```suggestion
```

File: context/package-lock.json [hunk 4]
```diff
@@ -15694,6 +15694,11 @@
       "integrity": "sha512-mT8iDcrh03qDGRRmoA2hmBJnxpllMR+0/0qlzjqZES6NdiWDcZkCNAk4rPFZ9Q85r27unkiNNg8ZOiwZXBHwcA==",
       "dev": true
     },
+    "chart.js": {
+      "version": "3.4.1",
+      "resolved": "https://registry.npmjs.org/chart.js/-/chart.js-3.4.1.tgz",
+      "integrity": "sha512-0R4mL7WiBcYoazIhrzSYnWcOw6RmrRn7Q4nKZNsBQZCBrlkZKodQbfeojCCo8eETPRCs1ZNTsAcZhIfyhyP61g=="
+    },
```
Comment (mccalluc) [inline context/package-lock.json:15701]: Just to keep things tidy, I'd say revert this now, and we can bring it back when you need it. It's good to keep PRs on topic, and not let unrelated changes creep in.
```suggestion
```
... or, alternatively, add the charting to this PR... but I would lean towards keeping it small.

**Review by mccalluc — CHANGES_REQUESTED** [review]

File: context/app/static/js/pages/Vis/Vis.jsx [hunk 5]
```diff
@@ -33,55 +33,76 @@ import useSearchData from 'js/hooks/useSearchData';
...
+          },
+          {
+            'mapped_metadata.race': {
+              terms: {
+                field: 'mapped_metadata.race.keyword',
+              },
+            },
+          },
+        ],
+        size: 10000,
```
[the remark this answers is not in the document]
Comment (john-conroy) [reply context/app/static/js/pages/Vis/Vis.jsx:57]: Actually, for these composite queries the inner size param needs to be set to non 0.

File: context/package.json (hunk 2 above)
Comment (john-conroy) [reply context/package.json:21]: Alternatively you can just run `npm uninstall chart.js` and it will uninstall the package.

File: context/app/static/js/pages/Vis/Vis.jsx [hunk 6]
```diff
@@ -33,55 +33,94 @@ import useSearchData from 'js/hooks/useSearchData';
...
-  const { buckets } = searchData.aggregations['mapped_metadata.race'];
+
+  function getCount(buckets, bloodType, race) {
+    const filtered = buckets.filter(
+      (b) => b.key['mapped_metadata.blood_type'] === bloodType && b.key['mapped_metadata.race'] === race,
+    );
+    return filtered.length ? filtered[0].doc_count : 0;
+  }
+  function getKeyValues(buckets, key) {
+    return [...new Set(buckets.map((b) => b.key[key]))];
```
Comment (mccalluc) [inline context/app/static/js/pages/Vis/Vis.jsx:78]:
```suggestion
    return new Set(buckets.map((b) => b.key[key]));
```
This came up in the call: I think you can just return a set here, instead of converting it back to a list... but I could be wrong.

File: context/app/static/js/pages/Vis/Vis.jsx [hunk 7]
```diff
@@ -33,55 +33,94 @@ import useSearchData from 'js/hooks/useSearchData';
...
+    Object.keys(searchData).length && (
+      <Paper>
+        <StyledTableContainer>
+          <Table stickyHeader>
+            <TableHead>
+              <TableRow>
+                <HeaderCell> </HeaderCell>
+                {getKeyValues(buckets, 'mapped_metadata.blood_type').map((type) => (
+                  <HeaderCell> {type} </HeaderCell>
+                ))}
```
Comment (mccalluc) [inline context/app/static/js/pages/Vis/Vis.jsx:92]: This is good, but you can't be certain the order coming out of this function will be the same as what you list below. I'd suggest taking this in stages:
- In this PR, replace this `map` with an explicit list.
- After it's merged, come back and add `const bloodTypes = getKeyValues(buckets, 'mapped_metadata.blood_type')` at the top, and use it here, and in the rows below.
- and finally wrap the rows on the table in a similar loop for race.

**Review by mccalluc — CHANGES_REQUESTED** [review]
See comment: For now, replace `map` with an explicit list. You can clean it up later, but we want to be very sure we never introduce code that could just give us the wrong information.

File: context/app/static/js/pages/Vis/Vis.jsx (hunk 6 above)
Comment (Roselkisma) [reply context/app/static/js/pages/Vis/Vis.jsx:78]: You are right, we talked about it, I forgot to take it out.

**Review by mccalluc — APPROVED** [review]
